docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,125 @@
+
+Another alternative is to implement more tiers, or user-specified numerical priorities that guide Kubelet on which tasks to kill (if jobs are running out of memory, lower priority tasks will be killed). For now, these alternatives will not be implemented because they don’t seem to add much value, could lead to a more cluttered user interface, and are more difficult to implement. Many of these functionalities either involve racy user-space control loops, or kernel modifications.
+
+## Implementation
+
+Implement QoS policy
+- Add a function in kubelet.go that decides whether a pod is best-effort or top-tier
+- For now, return based on whether container limits are specified, but eventually this can be refactored based on a user specified value in the pod spec
+
+Change memory overcommit mode
+- Right now overcommit mode is off on the machines we set up, so if there isn’t enough memory malloc will return null. This prevents QoS, because best-effort pods won’t be evicted. Instead, when there isn’t enough memory, and top-tier pods call malloc, they may not get the memory they want. We want top-tier pods to get the memory they request, and force out best-effort pods.
+- Change the memory overcommit mode to 1, so that using excess memory starts the OOM killer.
+
+Pod admission
+- Modify predicates.CheckPodsExceedingCapacity (in scheduler/algorithm/predicates, which is used by both the Kubelet and the Scheduler)
+- If the pod is best-effort, we admit it automatically
+- Otherwise, admit the top-tier pod if there’s space based on limits of the other top-tier pods
+
+Pod configuration
+- We’re focusing on Docker in this implementation (not RKT)
+- OOM scores
+  - Note that the OOM score of a process is 10 times the % of memory the process consumes, adjusted by OOM_SCORE_ADJ, barring exceptions (e.g. process is launched by different users). Processes with higher OOM scores are killed.
+  - The base OOM score is between 0 and 1000, so if process A’s OOM_SCORE_ADJ - process B’s OOM_SCORE_ADJ is over a 1000, then process A will die.
+- Best-effort
+  - In the initial implementation, ignore CPU limits, set CPU shares to 2. In the final implementation, create a “best-effort” cgroup that contains all the best-effort pods. The best-effort cgroup will get 2 CPU shares. Within the cgroup, each container will be in a cgroup that gets CPU shares proportional to the amount of CPU it request for (or 2 shares if no limit was specified).
+  - Set memory limits as specified by the user
+  - Set OOM_SCORE_ADJ to 501
+- Top-tier
+  - Set CPU shares and memory requirements as we do right now
+  - Set OOM_SCORE_ADJ to -501

```

Comment:
That's a good point. There would be an issue if the user launches lots of top-tier pods that don't take up much memory. Then all the top-tier pods, the Kubelet, and the docker daemon would have the same oom_score of 0, and we cannot guarantee what gets evicted. Is this the issue that you're concerned about?

I think we could avoid this by keeping some padding for the kubelet, OS, etc (which we don't do at the moment).